### PR TITLE
[REVIEW] Add new factory methods to include passing an existing null mask

### LINKED ISSUE #3235: [FEA] Add column factories the allow passing in an existing null mask
**Is your feature request related to a problem? Please describe.**

There is currently no convenient way to create a new column using an existing bitmask (either copying or moving). 

**Describe the solution you'd like**

Add `make_numeric_column` and `make_timestamp_column` that accept a universal reference for an existing bitmask.

### PR DESCRIPTION
**Pull request by davidwendt**
Closes #3235

I have many functions where the output column's null mask should be the same (copy of) the input column's null mask. So the code just copies the null mask before calling the `column` constructor like this:
```
// copy null mask
rmm::device_buffer null_mask;
cudf::size_type null_count = input_column.null_count();
if( input_column.has_nulls() )
    null_mask = rmm::device_buffer( input_column.null_mask(),
                                    bitmask_allocation_size_bytes(input_column.size()),
                                    stream, mr);
// create output column
auto output_column = std::make_unique<cudf::column>( output_type, input_column.size(),
    rmm::device_buffer(input_column.size() * size_of(output_type), stream, mr),
    null_mask, null_count);
```
This PR will add new column factory methods to pass an existing null mask when building the column so the above code can be reduced to the following:
```
auto output_column = make_numeric_column( output_type, input_column.size(),
                                          input_column.null_mask() );
```

### CODE REVIEW

**Comment by jrhemstad** [comment]
This API would be more generic if it accept a universal reference to a device buffer: 
```
template <typename Buffer>
std::unique_ptr<column> make_numeric_column(
    data_type type, size_type size, Buffer&& null_mask,
    size_type null_count = cudf::UNKNOWN_NULL_COUNT, cudaStream_t stream = 0,
    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
```

Which allows the following uses:
```
make_numeric_column(..., rmm::device_buffer(p, bitmask_allocation_size_bytes(size)), ...); // deep copies the mask from the raw pointer p

rmm::device_buffer null_mask = ....;
make_numeric_column(..., null_mask, ...); // deep copies the mask from the device_buffer null_mask

make_numeric_column(..., std::move(null_mask), ...); // moves contents of `null_mask` into new column
```

**Comment by davidwendt** [comment]
So I don't want to have to know how to build a `device_buffer` from a `bitmask_t*` (returned from `column::null_mask()`) which requires knowing about `bitmask_allocation_size_bytes` and what include file that is in, etc. Also, if the `bitmask_t*` is null then an if-statement is required otherwise you get a "Invalid size" exception from the `device_buffer` constructor so you would need to know to pass 0 for the size instead.
I'd rather the factory just handle all these details.
Perhaps another factory method is needed in case you have a `device_buffer` of a null mask. Otherwise, creating a proper `device_buffer` from a `column::null_mask()` is still about the same amount of work that is in the description.

File: cpp/include/cudf/column/column_factories.hpp [hunk 1]
```diff
@@ -42,13 +42,37 @@ namespace cudf
  * kernels
  * @param[in] mr Optional resource to use for device memory
  * allocation of the column's `data` and `null_mask`.
- *---------------------------------------------------------------------------**/
```
Comment (harrism) [inline cpp/include/cudf/column/column_factories.hpp]: I love how @jrhemstad keeps adding horizontal rules and @davidwendt keeps removing 'em. :)

Comment (jrhemstad) [reply cpp/include/cudf/column/column_factories.hpp]: It's not something I usually think about. My VSCode is just set up to add them. I've changed the setting now.

File: cpp/src/column/column_factories.cpp [hunk 2]
```diff
@@ -52,6 +52,25 @@ std::unique_ptr<column> make_numeric_column(
...
+
+  // copy null mask
+  rmm::device_buffer copy_null_mask;
+  if( null_mask != nullptr )
+      copy_null_mask = rmm::device_buffer( null_mask,
+                                           bitmask_allocation_size_bytes(size),
+                                           stream, mr);
+  return std::make_unique<column>(
+      type, size, rmm::device_buffer{size * cudf::size_of(type), stream, mr},
+      copy_null_mask, null_count,
```
Comment (jrhemstad) [inline cpp/src/column/column_factories.cpp]: Need to be careful about copies vs. moves. This was performing an extra copy here. 
```suggestion
      std::move(copy_null_mask), null_count,
```

File: cpp/src/column/column_factories.cpp [hunk 3]
```diff
@@ -52,6 +52,25 @@ std::unique_ptr<column> make_numeric_column(
...
+      std::vector<std::unique_ptr<column>>{});
```
Comment (jrhemstad) [inline cpp/src/column/column_factories.cpp]: This is already defaulted in the `column` ctor, so no need to have this. 
```suggestion
      );
```

File: cpp/src/column/column_factories.cpp [hunk 4]
```diff
@@ -64,4 +83,23 @@ std::unique_ptr<column> make_timestamp_column(
...
+
+  // copy null mask
+  rmm::device_buffer copy_null_mask;
+  if( null_mask != nullptr )
+      copy_null_mask = rmm::device_buffer( null_mask,
+                                           bitmask_allocation_size_bytes(size),
+                                           stream, mr);
+  return std::make_unique<column>(
+      type, size, rmm::device_buffer{size * cudf::size_of(type), stream, mr},
+      copy_null_mask, null_count,
```
Comment (jrhemstad) [inline cpp/src/column/column_factories.cpp]:
```suggestion
      std::move(copy_null_mask), null_count,
```

File: cpp/src/column/column_factories.cpp [hunk 5]
```diff
@@ -64,4 +83,23 @@ std::unique_ptr<column> make_timestamp_column(
...
+      std::vector<std::unique_ptr<column>>{});
```
Comment (jrhemstad) [inline cpp/src/column/column_factories.cpp]:
```suggestion
      );
```

**Comment by davidwendt** [comment]
I guess I like this option the best then.
```
make_numeric_column(..., copy_bitmask(view.null_mask(), view.offset(), view.size()), ...);
```
Could I have a `copy_bitmask(column_view)` method in #3262 as well which takes care of these parameters?

**Comment by jrhemstad** [comment]
> Could I have a `copy_bitmask(column_view)` method in #3262 as well which takes care of these parameters?

I like this. It should be a trivial extension to the other `copy_bitmask` API.

File: cpp/include/cudf/column/column_factories.hpp (hunk 1 above)
Comment (harrism) [reply cpp/include/cudf/column/column_factories.hpp]: I'm just teasing you -- I don't care either way. It was not I who complained about the lines in the past...

File: cpp/include/cudf/column/column_factories.hpp [hunk 6]
```diff
@@ -42,13 +42,38 @@ namespace cudf
...
+ * @param[in] null_mask Null mask to use for this column.
+ * @param[in] null_count Optional number of nulls in the null_mask.
+ * @param[in] stream Optional stream on which to issue all memory allocation and device
+ * kernels
+ * @param[in] mr Optional resource to use for device memory
+ * allocation of the column's `data` and `null_mask`.
+ */
+std::unique_ptr<column> make_numeric_column(
+    data_type type, size_type size, 
+    rmm::device_buffer&& null_mask,
```
Comment (jrhemstad) [inline cpp/include/cudf/column/column_factories.hpp]: This should be a universal reference to allow copying _or_ moving.

Like this:
```
template <typename B>
std::unique_ptr<column> make_numeric_column(
    data_type type, size_type size, 
    B&& null_mask,size_type null_count = cudf::UNKNOWN_NULL_COUNT, cudaStream_t stream = 0,
    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());

File: cpp/include/cudf/column/column_factories.hpp [hunk 7]
```diff
@@ -66,13 +91,38 @@ std::unique_ptr<column> make_numeric_column(
...
+ * @param[in] null_mask Null mask to use for this column.
+ * @param[in] null_count Optional number of nulls in the null_mask.
+ * @param[in] stream Optional stream on which to issue all memory allocation and device
+ * kernels
+ * @param[in] mr Optional resource to use for device memory
+ * allocation of the column's `data` and `null_mask`.
+ */
+std::unique_ptr<column> make_timestamp_column(
+    data_type type, size_type size,
+    rmm::device_buffer&& null_mask,
```
Comment (jrhemstad) [inline cpp/include/cudf/column/column_factories.hpp]: Universal ref.

File: cpp/src/column/column_factories.cpp [hunk 8]
```diff
@@ -48,8 +48,20 @@ std::unique_ptr<column> make_numeric_column(
...
+std::unique_ptr<column> make_numeric_column(
+    data_type type, size_type size,
+    rmm::device_buffer&& null_mask,
+    size_type null_count,
+    cudaStream_t stream, rmm::mr::device_memory_resource* mr) {
+  CUDF_EXPECTS(is_numeric(type), "Invalid, non-numeric type.");
+
+  return std::make_unique<column>(
+      type, size, rmm::device_buffer{size * cudf::size_of(type), stream, mr},
+      null_mask, null_count);
```
Comment (jrhemstad) [inline cpp/src/column/column_factories.cpp]:
```suggestion
      std::forward(null_mask), null_count);
```

File: cpp/src/column/column_factories.cpp [hunk 9]
```diff
@@ -64,4 +76,17 @@ std::unique_ptr<column> make_timestamp_column(
...
+std::unique_ptr<column> make_timestamp_column(
+    data_type type, size_type size,
+    rmm::device_buffer&& null_mask,
+    size_type null_count,
+    cudaStream_t stream, rmm::mr::device_memory_resource* mr) {
+  CUDF_EXPECTS(is_timestamp(type), "Invalid, non-timestamp type.");
+
+  return std::make_unique<column>(
+      type, size, rmm::device_buffer{size * cudf::size_of(type), stream, mr},
+      null_mask, null_count);
```
Comment (jrhemstad) [inline cpp/src/column/column_factories.cpp]:
```suggestion
      std::forward(null_mask), null_count);
```

File: cpp/include/cudf/column/column_factories.hpp (hunk 6 above)
Comment (jrhemstad) [reply cpp/include/cudf/column/column_factories.hpp]: https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers

File: cpp/include/cudf/column/column_factories.hpp [hunk 10]
```diff
@@ -42,13 +44,45 @@ namespace cudf
...
+std::unique_ptr<column> make_numeric_column(
+    data_type type, size_type size, 
+    B&& null_mask,
+    size_type null_count = cudf::UNKNOWN_NULL_COUNT, cudaStream_t stream = 0,
+    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource())
+{
+  CUDF_EXPECTS(is_numeric(type), "Invalid, non-numeric type.");
+  return std::make_unique<column>(
+      type, size, rmm::device_buffer{size * cudf::size_of(type), stream, mr},
+      std::forward(null_mask), null_count);
```
Comment (devavret) [inline cpp/include/cudf/column/column_factories.hpp]: Template argument deduction is disabled with `std::forward`
```suggestion
      std::forward<B>(null_mask), null_count);
```

File: cpp/include/cudf/column/column_factories.hpp [hunk 11]
```diff
@@ -66,13 +100,45 @@ std::unique_ptr<column> make_numeric_column(
...
+std::unique_ptr<column> make_timestamp_column(
+    data_type type, size_type size,
+    B&& null_mask,
+    size_type null_count = cudf::UNKNOWN_NULL_COUNT, cudaStream_t stream = 0,
+    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource())
+{
+  CUDF_EXPECTS(is_timestamp(type), "Invalid, non-timestamp type.");
+  return std::make_unique<column>(
+      type, size, rmm::device_buffer{size * cudf::size_of(type), stream, mr},
+      std::forward(null_mask), null_count);
```
Comment (devavret) [inline cpp/include/cudf/column/column_factories.hpp]:
```suggestion
      std::forward<B>(null_mask), null_count);
```

**Pull request merged**